Tools/autotest/common.py

```diff
@@ -2339,137 +2339,147 @@ def delay_sim_time(self, seconds_to_wait):
+
+        def get_groundspeed(timeout2):
+            msg = self.mav.recv_match(type='VFR_HUD', blocking=True, timeout=timeout2)
+            if msg:
+                return msg.groundspeed
+
+        def validator(value2, target2=None):
+            if speed_min <= value2 <= speed_max:
                 return True
-        raise WaitGroundSpeedTimeout("Failed to attain groundspeed range")
+            else:
+                return False
+
+        self.wait_and_maintain(value_name="Groundspeed", target=speed_min, achieved_getter=lambda: get_groundspeed(timeout), accuracy=(speed_max - speed_min), validator=lambda value2, target2: validator(value2, target2), timeout=timeout, **kwargs)
 
-    def wait_roll(self, roll, accuracy, timeout=30):
+    def wait_roll(self, roll, accuracy, timeout=30, **kwargs):
         """Wait for a given roll in degrees."""
-        tstart = self.get_sim_time()
-        self.progress("Waiting for roll of %d at %s" % (roll, time.ctime()))
-        while self.get_sim_time_cached() < tstart + timeout:
-            m = self.mav.recv_match(type='ATTITUDE', blocking=True)
-            p = math.degrees(m.pitch)
-            r = math.degrees(m.roll)
-            self.progress("Roll %d Pitch %d" % (r, p))
-            if math.fabs(r - roll) <= accuracy:
-                self.progress("Attained roll %d" % roll)
-                return True
-        raise WaitRollTimeout("Failed to attain roll %d" % roll)
+        def get_roll(timeout2):
+            msg = self.mav.recv_match(type='ATTITUDE', blocking=True, timeout=timeout2)
+            if msg:
+                p = math.degrees(msg.pitch)
+                r = math.degrees(msg.roll)
+                self.progress("Roll %d Pitch %d" % (r, p))
+                return r
+
+        def validator(value2, target2):
+            return math.fabs((value2 - target2 + 180) % 360 - 180) <= accuracy
+
+        self.wait_and_maintain(value_name="Roll", target=roll, achieved_getter=lambda: get_roll(timeout), validator=lambda value2, target2: validator(value2, target2), accuracy=accuracy, timeout=timeout, **kwargs)
 
-    def wait_pitch(self, pitch, accuracy, timeout=30):
+    def wait_pitch(self, pitch, accuracy, timeout=30, **kwargs):
         """Wait for a given pitch in degrees."""
+        def get_pitch(timeout2):
+            msg = self.mav.recv_match(type='ATTITUDE', blocking=True, timeout=timeout2)
+            if msg:
+                p = math.degrees(msg.pitch)
+                r = math.degrees(msg.roll)
+                self.progress("Pitch %d Roll %d" % (p, r))
+                return p
+
+        def validator(value2, target2):
+            return math.fabs((value2 - target2 + 180) % 360 - 180) <= accuracy
+
+        self.wait_and_maintain(value_name="Pitch", target=pitch, achieved_getter=lambda: get_pitch(timeout), validator=lambda value2, target2: validator(value2, target2), accuracy=accuracy, timeout=timeout, **kwargs)
```

Comment:
I wonder if we should have "current_value_getter" or something rather than "achieved_getter" which might be considered a boolean have-we-succeeded sort of thing.

Reply:
done